src/api/cvc4cpp.cpp

```diff
@@ -2010,31 +1917,72 @@ Term Solver::mkConst(Kind kind, uint32_t arg1, uint32_t arg2, Term arg3) const
 
 Term Solver::mkVar(const std::string& symbol, Sort sort) const
 {
-  // CHECK: sort exists?
   return d_exprMgr->mkVar(symbol, *sort.d_type);
 }
 
 Term Solver::mkVar(Sort sort) const
 {
-  // CHECK: sort exists?
   return d_exprMgr->mkVar(*sort.d_type);
 }
 
 Term Solver::mkBoundVar(const std::string& symbol, Sort sort) const
 {
-  // CHECK: sort exists?
   return d_exprMgr->mkBoundVar(symbol, *sort.d_type);
 }
 
 Term Solver::mkBoundVar(Sort sort) const
 {
-  // CHECK: sort exists?
   return d_exprMgr->mkBoundVar(*sort.d_type);
 }
 
 /* Create terms                                                               */
 /* -------------------------------------------------------------------------- */
 
+void Solver::check_mk_term(Kind kind, uint32_t nchildren) const
+{
+  PrettyCheckArgument(isDefinedKind(kind),
+                      kind,
+                      "Invalid kind '%s'",
+                      kindToString(kind).c_str());
+  const CVC4::kind::MetaKind mk = kind::metaKindOf(extToIntKind(kind));
+  PrettyCheckArgument(
+      mk == kind::metakind::PARAMETERIZED || mk == kind::metakind::OPERATOR,
+      kind,
+      "Invalid kind '%s', "
+      "only operator-style terms are created with mkTerm(), "
+      "to create variables and constants see mkVar(), mkBoundVar(), "
+      "and mkConst().",
+      kindToString(kind).c_str());
+  if (nchildren)
+  {
+    const uint32_t n =
+        nchildren - (mk == CVC4::kind::metakind::PARAMETERIZED ? 1 : 0);
+    PrettyCheckArgument(
+        n >= minArity(kind) && n <= maxArity(kind),
+        kind,
+        "Terms with kind %s must have at least %u children and "
+        "at most %u children (the one under construction has %u)",
+        kindToString(kind).c_str(),
+        minArity(kind),
+        maxArity(kind),
+        n);
+  }
+}
+
+void Solver::check_mk_op_term(OpTerm opTerm, uint32_t nchildren) const
+{
+  const Kind kind = opTerm.getKind();
+  const CVC4::Kind int_kind = extToIntKind(kind);
+  const CVC4::Kind int_op_kind =
+      NodeManager::operatorToKind(opTerm.d_expr->getNode());
+  PrettyCheckArgument(int_kind != kind::BUILTIN
+                          && CVC4::kind::metaKindOf(int_op_kind)
+                                 != kind::metakind::PARAMETERIZED,
+                      opTerm,
+                      "This term constructor is for parameterized kinds only");
+  check_mk_term(kind, nchildren);
```

Comment:
This probably won't work out because `kind` is parametrized and `check_mk_term` does not accept parametrized kinds?

Reply:
You're right, I will have to duplicate the checks.